packages/aws-cdk-lib/aws-route53/lib/record-set.ts

```diff
@@ -406,8 +406,15 @@ export class RecordSet extends Resource implements IRecordSet {
     }
 
     if (this.weight !== undefined) {
-      const idPrefix = `WEIGHT_${this.weight}_ID_`;
-      return this.createIdentifier(idPrefix);
+      if (Token.isUnresolved(this.weight)) {
+        const replacement = 'XXX'; // XXX simply because 255 is the highest value for a record weight
+        const idPrefix = `WEIGHT_${replacement}_ID_`;
+        const idTemplate = this.createIdentifier(idPrefix);
+        return idTemplate.replace(replacement, Token.asString(this.weight));
```

Comment:
```suggestion
        const idPrefix = `WEIGHT_${Tokenization.stringifyNumber(this.weight)}_ID_`;
        return this.createIdentifier(idPrefix);
```

Reply:
Hi @GavinZZ , I am not sure if we can say that the length will be consistent by using `Tokenization.stringifyNumber`. Maybe you know more the internals than I do, but from what I can tell, this is basically like using `Token.asString` except it is type safe. So it is building a string like `${Token[TOKEN.0]}` and caching that using the `TokenMap` class.

The problem is that the numeric index part of that key is non-deterministic as well. I performed multiple runs of synthesis and always get a different value for that index. Each run, by chance, the number is always four digits in length, so I had values like `${Token[TOKEN.1032]}`, but from what I see in the code, we can't assume that will always be the case.

That numeric part is coming from `TokenMap`'s private field `tokenCounter` which is initialized as a random number. For each new token that is registered, the value is incremented. There could potentially be the case where our stringified token has one more or one fewer digits, just by chance.

This length could also be influenced by modifying code before the evaluation of our token. So suppose a developer adds code which is resolving a bunch of tokens before constructing their RecordSet (for example, adding and using a bunch more CfnParameters). Suddenly CDK may detect a change to the set ID because the length of the stringified token will be different, and thus the generated set ID will be shortened.

So it's for this reason that I decided to spoof the weight value when generating the set ID. If the things I mentioned above don't actually raise any concerns, then we can go ahead apply your suggestion and merge after.

Reply:
You're right. This could technically result in some error case. I tested it out and in some cases, it does return less digits. I'm happy to approve your current solution.